proposals/large-ipld-dags.md

```diff
@@ -0,0 +1,117 @@
+
+This is only one 🔥 because there are plenty of useful sub-32GiB datasets and non-IPFS datasets.
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+🎯🎯🎯
+
+If we don't solve this now, users will likely store large DAGs any way they can (e.g., as raw
+blocks). We could end up with a lot of unfortunately structured data in Filecoin that's difficult to
+retrieve and work with, especially from IPFS.
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+??
+
+## Project definition
+#### Brief plan of attack
+
+1. Implement selector support in `lotus client deal`. 
+2. Implement selector support in `lotus client retrieve`.
+3. Support automatically splitting large dags into across deals in `lotus client retrieve`. 
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+All three of the above commands have been implemented.
+
+NOTE: stopping anywhere along the way will yield a useful result. As long as the first step is finished (selector support for `lotus client deal`), we'll be able to store large structured IPFS data on-chain.
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+1. Developers can easily store large directory trees on Filecoin.
+2. Developers can easily retrieve individual files from large datasets on Filecoin.
+3. Snapshots of English Wikipedia can be stored on Filecoin.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+The primary risk is that there may be a lack of demand to store large IPFS-formatted datasets in Filecoin. That is, users storing large datasets (> 32GiB) may all be using custom formats and may not care about IPFS files/directories, partial retrieval, etc.
+
+Another risk is that the IPLD selector language may be insufficient to describe useful selectors over IPFS data. It should be at least possible to _store_ DAG subsets using IPLD selectors, but we may need new selectors to, e.g., download individual IPFS files.
+
+#### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+1. Don't support datasets > 32GiB.
+2. Store large datasets as raw objects instead of IPFS files and accept the fact that these datasets
+   will be difficult to query/retrieve from IPFS.
+
+#### Dependencies/prerequisites
+<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->
+
+None.
```

Comment:
we should think about if this can be deferred or done in parallel with having the lotus client / market work using ipld-prime

Reply:
Just using ipld-prime doesn't get us much. I need to be able to (a) make a deal over a selector and (b) retrieve a selector.

Reply:
https://github.com/protocol/web3-dev-team/pull/27 is probably a dependency.